pkg/search/proxy/framework/plugins/karmada/karmada.go

```diff
@@ -0,0 +1,76 @@
+package karmada
+
+import (
+	"context"
+	"net/http"
+	"net/url"
+	"path"
+
+	restclient "k8s.io/client-go/rest"
+
+	"github.com/karmada-io/karmada/pkg/search/proxy/framework"
+	"github.com/karmada-io/karmada/pkg/util/proxy"
+)
+
+const (
+	order = 3000
+)
+
+// Karmada proxies requests to karmada control panel.
+// For non-resource requests, or resources are not defined in ResourceRegistry,
+// we redirect the requests to karmada apiserver.
+// Usually the request are
+// - api index, e.g.: `/api`, `/apis`
+// - to workload created in karmada controller panel, such as deployments and services.
+type Karmada struct {
+	ProxyLocation  *url.URL
+	ProxyTransport http.RoundTripper
+}
+
+var _ framework.Plugin = (*Karmada)(nil)
+
+// New creates an instance of Karmada
+func New(dep framework.PluginDependency) (framework.Plugin, error) {
+	location, err := url.Parse(dep.RestConfig.Host)
+	if err != nil {
+		return nil, err
+	}
+
+	transport, err := restclient.TransportFor(dep.RestConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Karmada{
+		ProxyLocation:  location,
+		ProxyTransport: transport,
+	}, nil
+}
+
+// Order implements Plugin
+func (p *Karmada) Order() int {
+	return order
+}
+
+// SupportRequest implements Plugin
+func (p *Karmada) SupportRequest(request framework.ProxyRequest) bool {
+	return true
```

Comment:
When select plugins in runtime/framework, this always return true, so it will always choose this plugin?

Reply:
This plugin's order is the last one. It's actually a fallback plugin. See @ikaven1024 previous comment.

Reply:
understand

Reply:
The plugin framework uses a variant of "Chain of Responsibility Pattern". 


From "Design Patterns Elements of Reusable Object-Oriented Software"

> Avoid coupling the sender of a request to its receiver by giving more than one object a chance to handle the request. Chain the receiving objects and pass the request along the chain until an object handles it.

So, there will be only one plugin selected.

This is different from "Pipes and Filters Pattern". "Pipes and Filters Pattern" will run multiple filters all together.

From "Pattern-Oriented Software Architecture Vol.1"

> The Pipes and Filters architectural pattern provides a structure for systems that process a stream of data. Each processing step is encapsulated in a filter component. Data is passed through pipes between adjacent filters. Recombining filters allows you to build families of related systems.